mlir/lib/Dialect/SCF/Utils/Utils.cpp

```diff
@@ -473,6 +496,185 @@ LogicalResult mlir::loopUnrollByFactor(
   return success();
 }
 
+/// Check if bounds of all inner loops are defined outside of `forOp`
+/// and return false if not.
+static bool areInnerBoundsInvariant(scf::ForOp forOp) {
+  auto walkResult = forOp.walk([&](scf::ForOp innerForOp) {
+    if (!forOp.isDefinedOutsideOfLoop(innerForOp.getLowerBound()) ||
+        !forOp.isDefinedOutsideOfLoop(innerForOp.getUpperBound()) ||
+        !forOp.isDefinedOutsideOfLoop(innerForOp.getStep()))
+      return WalkResult::interrupt();
+
+    return WalkResult::advance();
+  });
+  return !walkResult.wasInterrupted();
+}
+
+/// Unrolls and jams this loop by the specified factor.
+LogicalResult mlir::loopUnrollJamByFactor(scf::ForOp forOp,
```

Comment:
This has a rather large overlap with the affine version, can we minimize it somehow? Specifically, can the implementation be ported to operate on `LoopLikeOpInterface` instead of specific loops? It would then live with the interface. If not, hoist out common parts, possible with templating or callbacks, and put it in the interface or utils.

Reply:
I am not sure how to merge / minimize the between scf and affine. I added only partial support in `scf.for` where affine has a larger support.

Reply:
If it would introduce more complexity than remove, then let's not do it...

Reply:
I agree, thanks for the whole review!